Source/OIDTokenRequest.m

```diff
@@ -274,7 +274,24 @@ - (NSURLRequest *)URLRequest {
   NSMutableDictionary *httpHeaders = [[NSMutableDictionary alloc] init];
 
   if (_clientSecret) {
-    NSString *credentials = [NSString stringWithFormat:@"%@:%@", _clientID, _clientSecret];
+    // https://tools.ietf.org/html/rfc6749#appendix-B
+    // Note: Space character is percent encoded, rather than as "+" as it's a valid encoding,
+    // and less ambiguous (and we don't need the "+"-encoding for aesthetic reasons)
+    
+    // Starts with the standard URL-allowed character set.
+    NSMutableCharacterSet *allowedCharacters =
+        [[NSCharacterSet URLQueryAllowedCharacterSet] mutableCopy];
+    // Keep only alphanumeric characters.
+    [allowedCharacters formIntersectionWithCharacterSet:[NSCharacterSet alphanumericCharacterSet]];
```

Comment:
This pattern would not have occurred to me because I would have assumed that alphanumeric characters are a strict subset of the above.

I like what you have because it’s explicit - but I have a question for my own benefit.... is my assumption incorrect?

Reply:
That's what I thought at first but that's not true, alphanumericCharacterSet contains a lot of other alphanumeric characters (thousands of characters!!) that are not part of the US-ASCII characters.

alphanumericCharacterSet from Apple documentation:
A character set containing the characters in Unicode General Categories L*, M*, and N*.

Another way to do would be to call removeCharactersInString: on the first set by specifying special characters. Or defining our own set by including characters  from A...Z, a...z and 0...9.

Reply:
Awesome! Thanks for that!!

Reply:
I think it's a mistake to encode *every* non-alphanumeric character. This results in non-reserved characters like `.` being encoded. Some systems may not expect this to happen.  E.g. let's say I issue clients with alphanumerics, and `.` characters only (like a certain well-known IdP). Since I know I'm not using any reserved characters, my OAuth implementation could be lazy and *not* url-decode these values. Now it will break. 

Maybe we should just follow the [latest spec](https://url.spec.whatwg.org/#urlencoded-serializing) on `application/x-www-form-urlencoded` (linked from [here](https://www.w3.org/TR/html5/sec-forms.html#application-x-www-form-urlencoded-encoding-algorithm), including the treatment of `+`).

That spec defines the following serializer:
```
0x20 (SP)
Append U+002B (+) to output.
0x2A (*)
0x2D (-)
0x2E (.)
0x30 (0) to 0x39 (9)
0x41 (A) to 0x5A (Z)
0x5F (_)
0x61 (a) to 0x7A (z)
Append a code point whose value is byte to output.

Otherwise
Append byte, percent encoded, to output.
```

Implementing this logic in iOS with the charset methods you're using would result in logic like so:
1. Build a charset allowing: ` *-._A-Za-z0-9` (nb. space included). %-encode any char outside this range.
2. Substitute ` ` for `+`.

NB. I withdraw my comment about `+` since it's actually pretty clear on this front, both the W3C, and the OAuth2 spec. Space is encoded as `+`, and `+` gets percent-encoded.

Reply:
Thanks for your detailed explanation. To build the character set I followed instructions from https://www.w3.org/TR/html401/interact/forms.html#h-17.13.4.1 saying:

> Non-alphanumeric characters are replaced by `%HH', a percent sign and two hexadecimal digits representing the ASCII code of the character.

But maybe the definition of alphanumeric here is not the same as on iOS.

Reply:
Thinking about this further... Julien, could you add some unit tests to clarify the situation here?

Reply:
Yes I will!